### PR TITLE
Template context parameter to callbacks

### PR DESCRIPTION
**Pull request by jmwilson**
Rather than rely on C-style `void *context` parameter for callbacks, in C++ context parameters can be bound using `std::bind` or lamdba captures to create callback function signatures free of generic pointers.

Templated convenience wrappers remain to allow any context type to be bound to a matching callback function.

### CODE REVIEW

File: src/BackgroundPublish.h [hunk 1]
```diff
@@ -26,9 +26,14 @@
 #include "Particle.h"
 
 using publish_completed_cb_t = std::function<void(particle::Error status,
+    const char *event_name,
```
Comment (eberseth) [inline src/BackgroundPublish.h:29]: nit, `char*` instead of `char *`

File: src/BackgroundPublish.h [hunk 2]
```diff
@@ -72,16 +77,83 @@ class BackgroundPublish {
      * @param[in] flags PublishFlags type for the request
      * @param[in] priority priority of message. Lowest is highest priority, zero indexed
      * @param[in] cb callback on publish success or failure
+     *
+     * @return TRUE if request accepted, FALSE if not
+     */
+    bool publish(const char* name,
+                 const char* data = nullptr,
```
Comment (eberseth) [inline src/BackgroundPublish.h:84]: What is the use case for a publish without the `data` argument?

Comment (jmwilson) [reply src/BackgroundPublish.h:84]: Looks like it's for an event where the event name itself is the only information?

**Comment by eberseth** [comment]
Heads up that there are some errors in the unit test.
```bash
$ make
Scanning dependencies of target background-publish-test
[ 25%] Building CXX object CMakeFiles/background-publish-test.dir/test/test.cpp.o
In file included from /home/eric/projects/asset/background-publish/test/test.cpp:1:0:
/home/eric/projects/asset/background-publish/src/BackgroundPublish.h: In member function ‘bool BackgroundPublish<NumQueues>::publish(const char*, const char*, PublishFlags, std::size_t, publish_completed_cb_t)’:
/home/eric/projects/asset/background-publish/src/BackgroundPublish.h:329:29: error: ‘INVALID_STATE’ is not a member of ‘particle::Error’
         cb(particle::Error::INVALID_STATE, name, data);
                             ^~~~~~~~~~~~~
/home/eric/projects/asset/background-publish/src/BackgroundPublish.h:335:29: error: ‘INVALID_ARGUMENT’ is not a member of ‘particle::Error’
         cb(particle::Error::INVALID_ARGUMENT, name, data);
                             ^~~~~~~~~~~~~~~~
/home/eric/projects/asset/background-publish/src/BackgroundPublish.h:343:29: error: ‘BUSY’ is not a member of ‘particle::Error’
         cb(particle::Error::BUSY, name, data);
                             ^~~~
CMakeFiles/background-publish-test.dir/build.make:62: recipe for target 'CMakeFiles/background-publish-test.dir/test/test.cpp.o' failed
make[2]: *** [CMakeFiles/background-publish-test.dir/test/test.cpp.o] Error 1
CMakeFiles/Makefile2:67: recipe for target 'CMakeFiles/background-publish-test.dir/all' failed
make[1]: *** [CMakeFiles/background-publish-test.dir/all] Error 2
Makefile:94: recipe for target 'all' failed
make: *** [all] Error 2
```

File: src/BackgroundPublish.h (hunk 2 above)
Comment (eberseth) [reply src/BackgroundPublish.h:84]: yeah, I guess that is supported

File: src/BackgroundPublish.h (hunk 1 above)
Comment (jmwilson) [reply src/BackgroundPublish.h]: This seems all over the place in our code, so I'd wait until there's a style document and we can just run everything through `clang-format`

**Review by eberseth — APPROVED** [review]

**Pull request merged**